libs/scripts/personalization.js

```diff
@@ -0,0 +1,374 @@
+          variants[vn].commands.push(variantInfo);
+        } else {
+          console.log('Invalid action found: ', line);
+        }
+      });
+    });
+
+    config.variants = variants;
+    config.variantNames = variantNames;
+    return config;
+  } catch (e) {
+    console.log('error parsing personalization config:', e, experiences);
+  }
+  return null;
+}
+
+export async function replaceInner(path, element) {
+  if (!path || !element) return false;
+  let plainPath = path.endsWith('/') ? `${path}index` : path;
+  plainPath = plainPath.endsWith('.plain.html') ? plainPath : `${plainPath}.plain.html`;
+  try {
+    const resp = await fetch(plainPath);
+    if (!resp.ok) {
+      throw new Error('Invalid response', resp);
+    }
+    const html = await resp.text();
+    element.innerHTML = html;
+    return true;
+  } catch (e) {
+    console.log(`error loading experiment content: ${plainPath}`, e);
+  }
+  return false;
+}
+
+function getPersonalizationVariant(variantNames = [], variantLabel = null) {
+  const tagNames = Object.keys(PERSONALIZATION_TAGS);
+  // handle multiple variants that are space / comma delimited
+  const matchingVariant = variantNames.find((variant) => {
+    const names = variant.split(RE_SPACE_COMMA).filter(Boolean);
+    if (names.some((name) => (
+      name === variantLabel || (tagNames.includes(name) && PERSONALIZATION_TAGS[name]())))
+    ) {
+      return variant;
+    }
+    return false;
+  });
+  return matchingVariant;
+}
+
+async function fetchManifest(path) {
+  try {
+    const resp = await fetch(path);
+    if (!resp.ok) {
+      console.log('error loading manifest:', resp);
+      return null;
+    }
+    const json = await resp.json();
+    return json.data;
+  } catch (e) {
+    console.log(`error loading manifest: ${path}`, e);
+  }
+  return null;
+}
+
+export async function getPersConfig(name, variantLabel, manifestData, manifestPath) {
+  console.log('Personalization: ', name || manifestPath);
+
+  const data = manifestData || await fetchManifest(manifestPath);
+  const config = parseConfig(data);
+
+  if (!config) {
+    console.log('Error loading personalization config: ', name || manifestPath);
+    return {};
+  }
+
+  const selectedVariant = getPersonalizationVariant(config.variantNames, variantLabel);
+
+  if (selectedVariant && config.variantNames.includes(selectedVariant)) {
+    config.run = true;
+    config.selectedVariantName = selectedVariant;
+    config.selectedVariant = config.variants[selectedVariant];
+  }
+
+  config.name = name;
+  config.manifest = manifestPath;
+  return config;
+}
+
+export async function fragmentPersonalization(doc) {
+  const fpTableRows = doc.querySelectorAll('.fragment-personalization > div');
+  doc.querySelector('.fragment-personalization').remove();
+  if (!fpTableRows) return doc;
+
+  const variantNames = [];
+  const info = [...fpTableRows].reduce((obj, row) => {
+    if (row.children[0]?.innerText?.toLowerCase() === 'action') {
```

Comment:
Wdyt about giving the element names using destructuring instead of accessing the indexes?

So something like:
```js
  const info = [...fpTableRows].reduce((obj, row) => {
    const [actionEl, selectorEl, variantEl, fragment] = row.children
    if (actionEl?.innerText?.toLowerCase() === 'action') {
```